BondageClub/Scripts/ImageCache.js

```diff
@@ -0,0 +1,217 @@
+		/** @type string */
+		this.url = url;
+		/** @type CachedImageState */
+		this.state = CachedImageState.UNCACHED;
+		/** @type number */
+		this.errorCount = 0;
+		/** @type {function(CachedImage):void} */
+		this.loadCallback = loadCallback;
+		/** @type HTMLImageElement */
+		this.element = null;
+		/** @type {{ width: number; height: number; texture: WebGLTexture; }} */
+		this.textureInfo = null;
+	}
+
+	/**
+	 * Load the cached image if it's not already available
+	 */
+	load() {
+		if ((this.state == CachedImageState.LOADING) || (this.state == CachedImageState.LOADED))
+			return;
+
+		// Get a new image element to perform the loading for us
+		this.element = new Image();
+
+		// Set up listeners to keep track of the loading process
+		let image = this;
+		this.element.addEventListener("load", function() {
+			image.errorCount = 0;
+			image.state = CachedImageState.LOADED;
+			image.cache._imageStateDidChange(image);
+		});
+
+		this.element.addEventListener("error", function () {
+			image.errorCount += 1;
+			image.state = CachedImageState.ERRORED;
+			image.cache._imageStateDidChange(image);
+		});
+
+		// Start loading
+		this._refresh();
+	}
+
+	/**
+	 * Manually set the cached image to the given element.
+	 * @param {HTMLImageElement} element - The element to set
+	 */
+	setElement(element) {
+		this.element = element;
+		this.state = CachedImageState.LOADED;
+	}
+
+	/**
+	 * Refresh the given image.
+	 */
+	_refresh() {
+		this.state = CachedImageState.LOADING;
+		this.element.src = this.url;
+	}
+
+	/**
+	 * Is the image an asset image?
+	 * @returns boolean
+	 */
+	isAsset() { return (this.url.indexOf("Assets") >= 0); }
+
+	/**
+	 * Is the image loaded?
+	 */
+	isLoaded() { return this.state == CachedImageState.LOADED; }
+}
+
+/**
+ * The class responsible for loading and caching images
+ */
+class ImageCacheClass {
+	constructor() {
+		/** @type {Map<string, CachedImage>} */
+		this.cache = new Map();
+		this.loadedCount = 0;
+	}
+
+	/**
+	 * Get a cached image from the cache.
+	 *
+	 * @param {string} url - The URL of the image to lookup
+	 * @param {function(CachedImage):void} loadCallback - A callback that will be called when the load completes
+	 * @returns {CachedImage?} The cached image, or null
+	 */
+	get(url, loadCallback = null) {
+		var image = this.cache.get(url);
+		if (!image) {
+			image = new CachedImage(this, url, loadCallback);
+			this.cache.set(url, image);
+		}
+
+		// start loading the image
+		image.load();
+
+		// returns the final image
+		return image;
+	}
+
+	/**
+	 * Add a cached image to the cache.
+	 * @param {string} url - The URL of the image to store
+	 * @param {HTMLImageElement|HTMLCanvasElement} image - The element to use as the image
+	 */
+	set(url, image) {
+		var img = new CachedImage(this, url);
+		var realImage = null;
+		if (image instanceof HTMLCanvasElement) {
+			realImage = new Image();
+			realImage.src = image.toDataURL();
+		} else {
+			realImage = image;
+		}
```

Comment:
Could simplify this a little:
```suggestion
		let realImage = image;
		if (image instanceof HTMLCanvasElement) {
			realImage = new Image();
			realImage.src = image.toDataURL();
		}
```

Reply:
I did that because the (my? I'm using Atom) linter sees that a straight `let realImage = image` assignment could still be both types after the type check, and gives me a warning for it on `img.setElement(realImage)`. Arguably, the linter seems too clever for its own good 😉.

Reply:
Ah, I see. That'll be the passive type-checking in your IDE rather than the linter, but same effect. Happy to leave that as is in that case - it's not hurting anyone.

Reply:
@zorgjbe You can add the type manually above the variable:
```ts
/** @type {HTMLCanvasElement | Image} */
let realImage = image;
```
*(not sure if those exact types are what you want)*